libraries/pbes/include/mcrl2/pbes/tools/pbesiteration.h

```diff
@@ -0,0 +1,539 @@
+    if (x.name() == m_eq.variable().name())
+    {
+      data::mutable_indexed_substitution sigma;
+      data::data_expression_list pars = x.parameters();
+      for (const data::variable& v : m_eq.variable().parameters())
+      {
+        data::data_expression par = pars.front();
+        pars.pop_front();
+        if (par.sort() != v.sort())
+        {
+          // Parameters do not match with variables. Ignore this substitution.
+          result = x;
+          return;
+        }
+        sigma[v] = par;
+      }
+      pbes_expression p = pbes_rewrite(m_eq.formula(), m_pbes_rewriter, sigma);
+      std::set<propositional_variable_instantiation> set = find_propositional_variable_instantiations(p);
+      if (std::all_of(set.begin(),
+              set.end(),
+              [this](const propositional_variable_instantiation& v) { return v.name() != m_eq.variable().name(); }))
+      {
+        result = p;
+        m_stable = false;
+        return;
+      }
+      result = x;
+      return;
+    }
+    result = x;
+  }
+};
+
+void substitute(pbes_equation& into,
+    const pbes_equation& by,
+    substitute_propositional_variables_builder<pbes_system::pbes_expression_builder>& substituter)
+{
+  substituter.set_equation(by);
+  substituter.set_name(into.variable().name());
+  pbes_expression p;
+  substituter.apply(p, into.formula());
+  into.formula() = p;
+}
+
+data::data_expression pbestodata(pbes_equation& equation,
+    replace_propositional_variables_builder<pbes_system::pbes_expression_builder>& replace_substituter)
+{
+  pbes_expression expr;
+  replace_substituter.set_forward(true);
+  replace_substituter.set_name(equation.variable().name());
+  replace_substituter.apply(expr, equation.formula());
+
+  data::data_expression data_expr = atermpp::down_cast<data::data_expression>(detail::pbes2data(expr));
+
+  replace_substituter.set_forward(false);
+  replace_substituter.apply(expr, equation.formula());
+  replace_substituter.reset_instantiations();
+  return data_expr;
+}
+
+enum class InvResult
+{
+  INV_TRUE,
+  INV_FALSE,
+  EQ_FALSE
+};
+
+// Equation should be in full form.
+InvResult global_invariant_check(pbes_equation& equation,
+    substitute_propositional_variables_builder<pbes_system::pbes_expression_builder>& substituter,
+    replace_propositional_variables_builder<pbes_system::pbes_expression_builder>& replace_substituter,
+    data::data_specification data_spec,
+    simplify_data_rewriter<data::rewriter> pbes_rewriter,
+    std::set<data::variable>& global_variables)
+{
+  mCRL2log(log::verbose) << "INV: " << equation.symbol() << "\n";
+  if (equation.symbol().is_mu())
+  {
+    return InvResult::INV_FALSE;
+  }
+
+  // Base check: does it contain the right pvi
+  std::set<propositional_variable_instantiation> set = find_propositional_variable_instantiations(equation.formula());
+  if (set.size() == 0)
+  {
+    mCRL2log(log::verbose) << "INV: " << equation.variable().name() << " does not contain any pvi." << "\n";
```

Comment:
This is probably up to you, but perhaps moving some printing to the debug level (or even) trace might also be useful. We use verbose mostly for general progress messages.